src/components/CopyToClipboard/CopyToClipboard.tsx

```diff
@@ -7,75 +7,52 @@ export enum CopyToClipboardStatus {
     Error = 'error',
 }
 
-interface CopyToClipboardGeneralProps {
+export interface CopyToClipboardProps {
     children: (status: CopyToClipboardStatus) => React.ReactElement;
     text: string;
+    timeout?: number;
     onCopy?: (text: string, result: boolean) => void;
 }
 
-interface CopyToClipboardDefaultProps {
-    timeout: number;
-}
-
-interface CopyToClipboardInnerProps
-    extends CopyToClipboardGeneralProps,
-        CopyToClipboardDefaultProps {}
-
-export interface CopyToClipboardProps
-    extends CopyToClipboardGeneralProps,
-        Partial<CopyToClipboardDefaultProps> {}
-
-interface CopyToClipboardState {
-    status: CopyToClipboardStatus;
-}
-
-export class CopyToClipboard extends React.Component<
-    CopyToClipboardInnerProps,
-    CopyToClipboardState
-> {
-    static INITIAL_STATUS = CopyToClipboardStatus.Pending;
-
-    state: CopyToClipboardState = {
-        status: CopyToClipboard.INITIAL_STATUS,
-    };
-
-    private timerId?: number;
-
-    render() {
-        const {children, text} = this.props;
-        const {status} = this.state;
-        const content = children(status);
-
-        if (!React.isValidElement(content)) {
-            throw new Error('Content must be a valid react element');
-        }
-
-        return (
-            <ReactCopyToClipboard text={String(text)} onCopy={this.handleCopy}>
-                {content}
-            </ReactCopyToClipboard>
-        );
-    }
-
-    componentWillUnmount() {
-        clearTimeout(this.timerId);
+export function CopyToClipboard({children, text, timeout = 1000, onCopy}: CopyToClipboardProps) {
+    const resolvedText = React.useMemo(() => String(text), [text]);
+    const [status, setStatus] = React.useState(CopyToClipboard.INITIAL_STATUS);
+    const content = React.useMemo(() => children(status), [children, status]);
+    const timerIdRef = React.useRef<number>();
+    const handleCopy = React.useCallback<Required<ReactCopyToClipboard.Props>['onCopy']>(
+        (copyText, result) => {
+            setStatus(result ? CopyToClipboardStatus.Success : CopyToClipboardStatus.Error);
+
+            if (timerIdRef.current) {
+                clearTimeout(timerIdRef.current);
+            }
+            timerIdRef.current = window.setTimeout(() => {
```

Comment:
I guess it is would be more clear to have single effect on status change

```jsx
const onCopy = useCallback(() => {setStatus(OTHER_STATUS)}, [])

useEffect(() => {
  if (status === INITIAL_STATUS) {
    return;
  }

  const timer = setTimeout(() => {
    setStatus(INITIAL_STATUS);
  }, timeout);

  return () => {
    clearTimeout(timer);
  };
}, [status]);
```

Reply:
Moved this part of code to useEffect with status in deps